mm2src/coins/utxo/utxo_balance_events.rs

```diff
@@ -58,45 +58,29 @@ impl EventStreamer for UtxoBalanceEventStreamer {
         const RECEIVER_DROPPED_MSG: &str = "Receiver is dropped, which should never happen.";
         let streamer_id = self.streamer_id();
         let coin = self.coin;
+        let mut scripthash_to_address_map = BTreeMap::default();
 
-        async fn subscribe_to_addresses(
-            utxo: &UtxoCoinFields,
-            addresses: HashSet<Address>,
-        ) -> Result<BTreeMap<String, Address>, String> {
-            match utxo.rpc_client.clone() {
-                UtxoRpcClientEnum::Electrum(client) => {
-                    // Collect the scrpithash for every address into a map.
-                    let scripthash_to_address_map = addresses
-                        .into_iter()
-                        .map(|address| {
-                            let scripthash = address_to_scripthash(&address).map_err(|e| e.to_string())?;
-                            Ok((scripthash, address))
-                        })
-                        .collect::<Result<HashMap<String, Address>, String>>()?;
-                    // Add these subscriptions to the connection manager. It will choose whatever connections
-                    // it sees fit to subscribe each of these addresses to.
-                    client
-                        .connection_manager
-                        .add_subscriptions(&scripthash_to_address_map)
-                        .await;
-                    // Convert the hashmap back to btreemap.
-                    Ok(scripthash_to_address_map.into_iter().map(|(k, v)| (k, v)).collect())
-                },
-                UtxoRpcClientEnum::Native(_) => {
-                    Err("Balance streaming is currently not supported for native client.".to_owned())
-                },
-            }
-        }
-
-        if coin.as_ref().rpc_client.is_native() {
-            let msg = "Native RPC client is not supported for UtxoBalanceEventStreamer.";
-            ready_tx.send(Err(msg.to_string())).expect(RECEIVER_DROPPED_MSG);
-            panic!("{}", msg);
+        // Get all the addresses to subscribe to their balance updates.
+        let all_addresses = match coin.all_addresses().await {
+            Ok(addresses) => addresses,
+            Err(e) => {
+                let msg = format!("Failed to get all addresses: {e}");
+                ready_tx.send(Err(msg.clone())).expect(RECEIVER_DROPPED_MSG);
+                panic!("{}", msg);
+            },
+        };
+        // FIXME: This might take some good LONG time in an HD wallet with many addresses.
+        //        We better optimistically respond to the `ready_tx` to avoid blocking the enabler response to the RPC.
+        match subscribe_to_addresses(coin.as_ref(), all_addresses).await {
```

Comment:
on the otherhand, this call consists of multiple network requests (of how many addresses we register). this will slow down the streamer initialization, and subsequently the RPC response to the streamer enablement.

Reply:
Why not do this https://github.com/KomodoPlatform/komodo-defi-framework/blob/c6c66d74ff39ee0a82f2bd92469d31f1f6901335/mm2src/coins/utxo/rpc_clients/electrum_rpc/connection_manager/manager.rs#L232 concurrently instead?

Reply:
the thing is, we consider `ConnectionManager::add_subscriptions` infallible. It can go on forever and ever if the electrum connections are down at the moment and will not return till we are successful, and this will halt the `UtxoBalanceEventStreamer`, and that's OK during normal operation, but not OK during initialization.

making them subscribe concurrently is an overall optimization but doesn't affect the worst case scenario. this is why im more inclined towards optimistically returning `Ok` to the initializer.

Reply:
> the thing is, we consider ConnectionManager::add_subscriptions infallible

Does it have to be infallible?

Reply:
after writing this out and looking back at the `subscribe_to_addresses`. it really didn't need to be fallible in the first place.

it only fails when the rpc client is Native (which we can check in no-time beforehand) and also fails when `address_to_scripthash(address)` fails. i opted for gracefully just logging the error when `address_to_scripthash` fails - instead of failing to initialize the streamer (we also gracefully log it in `ScripthashNotification::SubscribeToAddresses`, so now we have it standardised more)

fixed here [b68cb02](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2431/commits/b68cb029b1ee12fafac8070d299da416ceb3bc47)

Reply:
> Does it have to be infallible?

if it is fallible, we would need to split the addresses into ones that we could successfully sub to and ones that we couldn't (within a set maximum trying time period) and we should try them at later time.

and later time means we wanna have a timer or something to remind us we should try again, and since this balance event streamer runs solely reactively (based on triggers of new balance events or an electrum disconnection so we have to re-sub some addresses), this kinda just adds complexity.

it is doable, but a little bit less pretty with no clear upsides compared to the current impl.